packages/twenty-server/src/modules/calendar/calendar-event-import-manager/drivers/microsoft-calendar/services/microsoft-calendar-get-events.service.ts

```diff
@@ -0,0 +1,91 @@
+import { Injectable } from '@nestjs/common';
+
+import {
+  PageCollection,
+  PageIterator,
+  PageIteratorCallback,
+} from '@microsoft/microsoft-graph-client';
+import { Event } from '@microsoft/microsoft-graph-types';
+
+import { formatMicrosoftCalendarEvents } from 'src/modules/calendar/calendar-event-import-manager/drivers/microsoft-calendar/utils/format-microsoft-calendar-event.util';
+import { parseMicrosoftCalendarError } from 'src/modules/calendar/calendar-event-import-manager/drivers/microsoft-calendar/utils/parse-microsoft-calendar-error.util';
+import { GetCalendarEventsResponse } from 'src/modules/calendar/calendar-event-import-manager/services/calendar-get-events.service';
+import { MicrosoftOAuth2ClientManagerService } from 'src/modules/connected-account/oauth2-client-manager/drivers/microsoft/microsoft-oauth2-client-manager.service';
+import { ConnectedAccountWorkspaceEntity } from 'src/modules/connected-account/standard-objects/connected-account.workspace-entity';
+
+@Injectable()
+export class MicrosoftCalendarGetEventsService {
+  constructor(
+    private readonly microsoftOAuth2ClientManagerService: MicrosoftOAuth2ClientManagerService,
+  ) {}
+
+  public async getCalendarEvents(
+    connectedAccount: Pick<
+      ConnectedAccountWorkspaceEntity,
+      'provider' | 'refreshToken' | 'id'
+    >,
+    syncCursor?: string,
+  ): Promise<GetCalendarEventsResponse> {
+    try {
+      const microsoftClient =
+        await this.microsoftOAuth2ClientManagerService.getOAuth2Client(
+          connectedAccount.refreshToken,
+        );
+
+      const { changedEvents, nextSyncToken } =
+        await this.getChangedCalendarEventIds(connectedAccount, syncCursor);
+
+      const events: Event[] = [];
+
+      for (const changedEvent of changedEvents) {
+        const event = await microsoftClient
+          .api(`/me/calendar/events/${changedEvent.id}`)
+          .get();
+
+        events.push(event);
+      }
```

Comment:
This is good for a first version. For the google synchronization we import all the events in one step, because all the information from the events is returned by the `.list` method. Here, and likely for other providers as well, we will need to do a two step import as it is done for messaging. We want to be able to first retrieve the event ids and to store them inside the redis cache, and then a second job `calendarImportJob` will be called to get the event content and save them. This allows us to be more robust if an error happens during the import.
Also, it could be great if we could do batch imports to get the content of multiple events in one request, but we can leave it up for a future version.

Reply:
I split the two calendar jobs into fetch and import. I can again import emails for microsoft - for google I would need to invest a bit of time and understand the google dev console to really test it :D.
I had to touch so many files I hope I didn't introduce any new bugs ... 

The batch request I would leave for later, from what I found you can just batch 20 requests which is less than I thought.